### ISSUE TITLE
Rework template include path to avoid clashes

### ISSUE THREAD

**Issue by codeinthehole** [django-oscar/django-oscar#1378]
The way Oscar handles templates can clash with other apps. For instance, when you use the `OSCAR_MAIN_TEMPLATE_DIR` folder in `TEMPLATE_DIRS` (as you need to do to install Oscar), then templates like `base.html` and `layout.html` will resolve to Oscar's template folder.  Buy other apps often define a `base.html` (eg Mezzanine) which will clash.

Oscar uses this additional template path to allow templates to be customised without forking the whole template.

I think we can work around this issue by changing how we use template roots. We should:
- Change `OSCAR_MAIN_TEMPLATE_DIR` to point to the `oscar` folder itself so the `base.html` template can be resolved against paths `oscar/base.html` and `templates/oscar/base.html` (but no longer `base.html`).
- Change all of Oscar's templates to extend the 'oscar'-namespaced version (eg `{% extends 'oscar/layout.html' %}` instead of `{% extends 'layout.html' %}`.

This will allow projects to customise Oscar's templates by extending the corresponding template but using the longer path:

``` html
# myproject/templates/oscar/base.html

{% extends 'templates/oscar/base.html' %}

{% block content %}

{% endblock %}
```

Previously, this customising template would have been located in `myproject/templates/base.html` and would have extended `oscar/base.html` and so this change will be backwards incompatible for projects that customise Oscar's templates.

**Comment by bxm156** [comment]
It would be great if oscar placed templates in their related apps where possible. For example

apps/checkout/templates/checkout/thank_you.html

Devs could then place their new templates in their forked apps, as well as use the existing django template path settings.

**Comment by specialunderwear** [comment]
I think that 'OSCAR_MAIN_TEMPLATE_DIR' should be removed alltogether, there is no reason to require users to manually specify the template path of an application, since ``django.template.loaders.app_directories.Loader`` is installed by default. All that need to be done to allow ``django.template.loaders.app_directories.Loader`` to find the oscar templates, is add 'oscar' to the end of the list ``get_core_apps`` returns.

Obviously, the template paths should be fixed everywhere to become 'oscar/dashboard/file.html' instead of 'dashboard/file.html' that would nicely namespace the templates adding robustness against conflicts.

For the mechanism for allowing same template inheritance, I would just reverse the mechanism. Make users add 'templates/oscar' to the template path for their secondary oscar template path. That way they can extend oscar/base.html like this:
```
# inside project/templates/oscar/dashboard/file.html
    {% extends 'dashboard/file.html' %}  # secondary template path has no
```
That kind of reintroduces the conflict possibility, but this time it is inside the user's project, and he can always descide to include dirname(dirname('templates/oscar') to the filesystem  searchpath allowing him to do same template inheritance as described above.

So after removing the need for manual configuration of oscar templates, and removing the need for 'OSCAR_MAIN_TEMPLATE_DIR' we should add a 'OSCAR_SECONDARY_TEMPLATE_DIR' var that users can use to implement same template inheritance. It would have exactly the same value as the 'OSCAR_MAIN_TEMPLATE_DIR' because we reversed the main and secundary template dirs.

To remain backwards compatible, OSCAR_MAIN_TEMPLATE_DIR should be deprecated, with a warning.  Explaining users should move their templates inside a folder named 'oscar' to make sure they can be found in the future. Each view could then look for the template in the secondary folder before looking in the main folder, but only if OSCAR_MAIN_TEMPLATE_DIR is still present in the DIRS setting. Somewhat similar how the django admin looks for templates in different locations.

The work can be done by me, but lets first agree on the implementation first.

**Comment by solarissmoke** [comment]
@specialunderwear I agree that the `OSCAR_MAIN_TEMPLATE_DIR` hack should be removed. I am not 100% sure I understand how you propose to maintain backward compatiblity, but would be fine with asking projects to tweak their `DIRS` setting if they need backward compatibility. If you have time to make a PR for this then it would be great to try and fix this in Oscar 2.

**Comment by solarissmoke** [comment]
Fixed in #3000.

**Comment by prysie** [comment]
Now someone just needs to update the getting started guide to reflect this
https://django-oscar.readthedocs.io/en/releases-1.6/internals/getting_started.html
